pyro/distributions/coalescent.py

```diff
@@ -0,0 +1,306 @@
+            batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self.__dict__.get("_validate_args")
+        return new
+
+    def log_prob(self, value):
+        """
+        Computes likelihood as in equations 7-8 of [3].
+
+        This has time complexity ``O(T + S N log(N))`` where ``T`` is the
+        number of time steps, ``N`` is the number of leaves, and ``S =
+        sample_shape.numel()`` is the number of samples of ``value``.
+
+        This is differentiable wrt ``rate_grid`` but neither ``leaf_times`` nor
+        ``value = coal_times``.
+
+        :param torch.Tensor value: A tensor of coalescent times. These denote
+            sets of size ``leaf_times.size(-1) - 1`` along the trailing
+            dimension and can be unordered along that dimension.
+        :returns: Likelihood ``p(coal_times | leaf_times, rate_grid)``
+        :rtype: torch.Tensor
+        """
+        if self._validate_args:
+            self._validate_sample(value)
+        coal_times = value
+        phylogeny = _make_phylogeny(self.leaf_times, coal_times)
+
+        # Compute survival factors for closed intervals.
+        cumsum = self._unbroadcasted_rate_grid.cumsum(-1)
+        cumsum = torch.nn.functional.pad(cumsum, (1, 0), value=0)
+        integral = _interpolate(cumsum, phylogeny.times[..., 1:])  # ignore the final lonely leaf
+        integral = integral[..., :-1] - integral[..., 1:]
+        integral = integral.clamp(min=torch.finfo(integral.dtype).tiny)  # avoid nan
+        log_prob = -(phylogeny.binomial[..., 1:-1] * integral).sum(-1)
+
+        # Compute density of coalescent events.
+        i = coal_times.floor().clamp(min=0).long()
+        rates = phylogeny.coal_binomial * _gather(self._unbroadcasted_rate_grid, -1, i)
+        log_prob = log_prob + safe_log(rates).sum(-1)
+
+        return log_prob
+
+
+def _gather(tensor, dim, index):
+    """
+    Like :func:`torch.gather` but broadcasts.
+    """
+    if dim != -1:
+        raise NotImplementedError
+    shape = broadcast_shape(tensor.shape[:-1], index.shape[:-1]) + (-1,)
+    tensor = tensor.expand(shape)
+    index = index.expand(shape)
+    return tensor.gather(dim, index)
+
+
+def _interpolate(array, x):
+    """
+    Continuously index into the rightmost dim of an array, linearly
+    interpolating between array values.
+    """
+    with torch.no_grad():
+        x0 = x.floor().clamp(min=0, max=array.size(-1) - 2)
+        x1 = x0 + 1
+    f0 = _gather(array, -1, x0.long())
+    f1 = _gather(array, -1, x1.long())
+    return f0 * (x1 - x) + f1 * (x - x0)
+
+
+def _weak_memoize(fn):
+    cache = {}
+
+    @functools.wraps(fn)
+    def memoized_fn(*args):
+        key = tuple(map(id, args))
+        if key not in cache:
+            cache[key] = fn(*args)
+            for arg in args:
+                weakref.finalize(arg, cache.pop, key, None)
+        return cache[key]
+
+    return memoized_fn
```

Comment:
@neerajprad could you review this code? I wonder if it is interacting poorly with PyTorch?

Reply:
No the failure occurs even without this, so this can't be the problem...